### DISCUSSION TITLE
Confused about documentation example

### DISCUSSION THREAD

**Discussion by R-seny** [facebookresearch/xformers#908] [Q&A] [answered]
In the documentation for Memory-efficient attention ([link to relevant documentation](https://facebookresearch.github.io/xformers/components/ops.html)) we have an "equivalent pytorch code" part.

In that code, to get raw attention scores, we are multiplying a tensor Q with the shape [Batch, SeqLen, NumHeads, Dimension]
and a tensor K.transpose(-2, -1) with the shape [Batch, SeqLen, Dimension, NumHeads]. 

As an (intermediate) result, we get a tensor of the size [Batch, SeqLen, NumHeads, NumHeads], over the last dimension of which we then do softmax and then multiply the result by V.

I am a bit confused about how this corresponds to traditional multihead attention calculation since it seems that we are essentially attending across attention heads here instead of between sequence elements within one head. I would appreciate it if somebody could help me understand this part.

**Comment by danthe3rd** [comment, marked as answer]
Hi,
Thanks for pointing that out! Historically, there was no head dimension when this doc was written, and then everything makes sense.
But you are right, with 4d inputs, the code is not equivalent. We need to fix this

**Reply by R-seny** [reply]
I see, thank you for the answer! Good to know I was not going crazy :)
